README.md
**:warning: This repository is no longer actively maintained.**

# Helix Resolve-Git-Ref

> An OpenWhisk Web Action that resolves a git reference (branch or tag) to the corresponding commit sha.

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-resolve-git-ref.svg)](https://codecov.io/gh/adobe/helix-resolve-git-ref)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-resolve-git-ref.svg)](https://circleci.com/gh/adobe/helix-resolve-git-ref)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-resolve-git-ref.svg)](https://github.com/adobe/helix-resolve-git-ref/blob/main/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-resolve-git-ref.svg)](https://github.com/adobe/helix-resolve-git-ref/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-resolve-git-ref.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-resolve-git-ref)
[![Renovate enabled](https://img.shields.io/badge/renovate-enabled-brightgreen.svg)](https://renovatebot.com/)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Installation

## Usage

```bash
curl "https://adobeioruntime.net/api/v1/web/helix/helix-services/resolve-git-ref@v1?owner=adobe&repo=helix-resolve-git-ref&ref=main"
```

For more, see the [API documentation](docs/API.md).

## Development

### Deploying Helix Resolve-Git-Ref

Deploying Helix Resolve-Git-Ref requires the `wsk` command line client, authenticated to a namespace of your choice. For Project Helix, we use the `helix` namespace.

All commits to `main` that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-services/resolve-git-ref@ci<num>` and tagged with the CI build number.




CHANGELOG.md
## [1.10.36](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.35...v1.10.36) (2021-06-13)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.17 ([#517](https://github.com/adobe/helix-resolve-git-ref/issues/517)) ([a3b12b6](https://github.com/adobe/helix-resolve-git-ref/commit/a3b12b60b57a583a1f1c81629367c0163ce1ee4c))

## [1.10.35](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.34...v1.10.35) (2021-06-13)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.1.8 ([#516](https://github.com/adobe/helix-resolve-git-ref/issues/516)) ([640b3bb](https://github.com/adobe/helix-resolve-git-ref/commit/640b3bb383d6a59f84a28ae34c654656f184ff5a))

## [1.10.34](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.33...v1.10.34) (2021-06-13)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.16 ([#515](https://github.com/adobe/helix-resolve-git-ref/issues/515)) ([1644824](https://github.com/adobe/helix-resolve-git-ref/commit/1644824258f653583876837b9cd37a686de67e3e))

## [1.10.33](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.32...v1.10.33) (2021-06-12)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.3 ([#512](https://github.com/adobe/helix-resolve-git-ref/issues/512)) ([ece2940](https://github.com/adobe/helix-resolve-git-ref/commit/ece2940b4c24e55d21de1cd7d1ab9352199e827a))

## [1.10.32](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.31...v1.10.32) (2021-06-11)


### Bug Fixes

* **deps:** update dependency @adobe/helix-fetch to v2.3.0 ([#511](https://github.com/adobe/helix-resolve-git-ref/issues/511)) ([0b5587c](https://github.com/adobe/helix-resolve-git-ref/commit/0b5587cc67449c58fccdcee5d990d062fa14600f))

## [1.10.31](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.30...v1.10.31) (2021-06-05)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.15 ([#509](https://github.com/adobe/helix-resolve-git-ref/issues/509)) ([b1a2651](https://github.com/adobe/helix-resolve-git-ref/commit/b1a265142a840785926ee9f963cc279013e7f4f2))

## [1.10.30](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.29...v1.10.30) (2021-06-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.14 ([#507](https://github.com/adobe/helix-resolve-git-ref/issues/507)) ([67aff32](https://github.com/adobe/helix-resolve-git-ref/commit/67aff329625b99cb2ecc401a9fe47fe204f4b44a))

## [1.10.29](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.28...v1.10.29) (2021-05-31)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.13 ([#503](https://github.com/adobe/helix-resolve-git-ref/issues/503)) ([8de1b63](https://github.com/adobe/helix-resolve-git-ref/commit/8de1b631e605b02ac8993fcdef3e28a3251a672f))

## [1.10.28](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.27...v1.10.28) (2021-05-31)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.1 ([#502](https://github.com/adobe/helix-resolve-git-ref/issues/502)) ([4899d38](https://github.com/adobe/helix-resolve-git-ref/commit/4899d38501b82a1a8e3867ec2de439454919ad97))

## [1.10.27](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.26...v1.10.27) (2021-05-29)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.12 ([#500](https://github.com/adobe/helix-resolve-git-ref/issues/500)) ([5a113e0](https://github.com/adobe/helix-resolve-git-ref/commit/5a113e00c824e687deba967e21da8c1c2820e3cb))

## [1.10.26](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.25...v1.10.26) (2021-05-27)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.11 ([#499](https://github.com/adobe/helix-resolve-git-ref/issues/499)) ([4b00a1c](https://github.com/adobe/helix-resolve-git-ref/commit/4b00a1c9b95b143843e57e15eb5701a843339314))

## [1.10.25](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.24...v1.10.25) (2021-05-25)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.10 ([#496](https://github.com/adobe/helix-resolve-git-ref/issues/496)) ([c50bbd7](https://github.com/adobe/helix-resolve-git-ref/commit/c50bbd7d87f2cfa054b38d3edc9d40b8f1bcb248))

## [1.10.24](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.23...v1.10.24) (2021-05-17)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal-logger to v1.0.4 ([#490](https://github.com/adobe/helix-resolve-git-ref/issues/490)) ([75aa593](https://github.com/adobe/helix-resolve-git-ref/commit/75aa59304a18d2d721a5aec5c8ff7dfa588b2b04))

## [1.10.23](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.22...v1.10.23) (2021-05-17)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.9 ([#488](https://github.com/adobe/helix-resolve-git-ref/issues/488)) ([185c0ac](https://github.com/adobe/helix-resolve-git-ref/commit/185c0ac738661c6a4fdd665bc422f8ac605ed870))

## [1.10.22](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.21...v1.10.22) (2021-05-11)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.8 ([#483](https://github.com/adobe/helix-resolve-git-ref/issues/483)) ([3dec771](https://github.com/adobe/helix-resolve-git-ref/commit/3dec771cc9618acf16d95c8001773af791c0900a))

## [1.10.21](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.20...v1.10.21) (2021-05-11)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared to v8.0.7 ([#482](https://github.com/adobe/helix-resolve-git-ref/issues/482)) ([3057d8b](https://github.com/adobe/helix-resolve-git-ref/commit/3057d8b8dc436ac08a4aa85e3b07b13c9f19234d))

## [1.10.20](https://github.com/adobe/helix-resolve-git-ref/compare/v1.10.19...v1.10.20) (2021-05-09)


### Bug Fixes


test/index.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

'use strict';

process.env.HELIX_FETCH_FORCE_HTTP1 = 'true';

const assert = require('assert');
const path = require('path');
const querystring = require('querystring');

const NodeHttpAdapter = require('@pollyjs/adapter-node-http');
const FSPersister = require('@pollyjs/persister-fs');
const { setupMocha: setupPolly } = require('@pollyjs/core');
const nock = require('nock');
const { fetch, reset } = require('@adobe/helix-fetch').context({ alpnProtocols: ['http/1.1'] });
const pkgJson = require('../package.json');

const OWNER = 'adobe';
const REPO = 'helix-resolve-git-ref';
const PRIVATE_REPO = 'project-helix';
const SHORT_REF = 'main';
const FULL_REF = 'refs/heads/main';

const index = require('../src/index');

function main(params, headers = {}, context = {}) {
  if (!context.env) {
    context.env = {};
  }
  // emulate universal api
  return index.main({
    url: `https://www.dummy.com/resolve?${querystring.stringify(params)}`,
    headers: new Map(Object.entries(headers)),
  }, context);
}

/**
 * Checks if the specified string is a valid SHA-1 value.
 *
 * @param {string} str
 * @returns {boolean} `true` if `str` represents a valid SHA-1, otherwise `false`
 */
function isValidSha(str) {
  if (typeof str === 'string' && str.length === 40) {
    const res = str.match(/[0-9a-f]/g);
    return res && res.length === 40;
  }
  return false;
}

describe('main tests', () => {
  setupPolly({
    recordFailedRequests: false,
    recordIfMissing: false,
    matchRequestsBy: {
      headers: {
        exclude: ['authorization', 'User-Agent', 'accept', 'accept-encoding'],
      },
    },
    logging: false,
    adapters: [NodeHttpAdapter],
    persister: FSPersister,
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, 'fixtures/recordings'),
      },
    },
  });

  it('main function is present', () => {
    assert(typeof main === 'function');
  });

  it('main function returns an object', async () => {
    const result = await main({});
    assert.equal(typeof result, 'object');
  });

  it('main function returns 400 for missing owner param', async () => {
    const { status } = await main({ repo: REPO, ref: SHORT_REF });
    assert.equal(status, 400);
  });

  it('main function returns 400 for missing repo param', async () => {
    const { status } = await main({ owner: OWNER, ref: SHORT_REF });
    assert.equal(status, 400);
  });

  it('ref param is optional (fallback: default branch)', async () => {
    const resp = await main({ owner: OWNER, repo: REPO });
    const { fqRef } = await resp.json();
    assert.equal(resp.status, 200);
    assert.equal(fqRef, 'refs/heads/main');
  });

  it('ref param is optional (fallback: default branch) for test repo', async () => {
    const resp = await main({ owner: 'trieloff', repo: 'test' });
    const { fqRef } = await resp.json();
    assert.equal(resp.status, 200);
    assert.equal(fqRef, 'refs/heads/main');
  });

  it('main function returns valid sha format', async () => {
    const resp = await main({ owner: OWNER, repo: REPO, ref: SHORT_REF });
    const { sha } = await resp.json();
    assert.equal(resp.status, 200);
    assert(isValidSha(sha));
  });

  it('main function supports short and full ref names', async () => {
    const resp1 = await main({ owner: OWNER, repo: REPO, ref: SHORT_REF });
    const resp2 = await main({ owner: OWNER, repo: REPO, ref: FULL_REF });
    const { sha: sha1 } = await resp1.json();
    const { sha: sha2 } = await resp2.json();
    assert.equal(sha1, sha2);
  });

  it('main function resolves tag', async () => {
    const ref = 'v1.0.0';
    let resp = await main({ owner: OWNER, repo: REPO, ref });
    const { fqRef, sha: sha1 } = await resp.json();
    assert.equal(fqRef, `refs/tags/${ref}`);
    resp = await main({ owner: OWNER, repo: REPO, ref: `refs/tags/${ref}` });
    const { sha: sha2 } = await resp.json();
    assert.equal(sha1, sha2);
  });

  it('main function returns correct sha', async () => {
    let resp = await main({ owner: OWNER, repo: REPO, ref: SHORT_REF });
    const { sha } = await resp.json();
    try {
      resp = await fetch(`https://api.github.com/repos/${OWNER}/${REPO}/branches/${SHORT_REF}`, { cache: 'no-store' });
      assert.ok(resp.ok);
      const { commit } = await resp.json();
      assert.ok(commit);
      assert.equal(commit.sha, sha);
    } finally {
      await reset();
    }
  });

  it('main function returns 404 for non-existing ref', async () => {
    const { status } = await main({ owner: OWNER, repo: REPO, ref: 'unknown' });
    assert.equal(status, 404);
  });

  it('main function returns 404 for non-existing repo', async () => {
    const { status } = await main({ owner: OWNER, repo: 'unknown', ref: SHORT_REF });
    assert.equal(status, 404);
  });

  it('main() works with private GitHub repo (gh token via header)', async () => {
    const resp = await main({
      owner: OWNER,
      repo: PRIVATE_REPO,
      ref: SHORT_REF,
    }, {
      'x-github-token': 'undisclosed-github-token',
    });
    const { fqRef } = await resp.json();
    assert.equal(resp.status, 200);
    assert.equal(fqRef, FULL_REF);
  });

  it('main() works with private GitHub repo (gh token via param)', async () => {
    const resp = await main({
      owner: OWNER,
      repo: PRIVATE_REPO,
      ref: SHORT_REF,
      GITHUB_TOKEN: 'undisclosed-github-token',
    });
    const { fqRef } = await resp.json();
    assert.equal(resp.status, 200);
    assert.equal(fqRef, FULL_REF);
  });

  it('main() works with private GitHub repo (gh token via env)', async () => {
    const resp = await main({
      owner: OWNER,
      repo: PRIVATE_REPO,
      ref: SHORT_REF,
    }, {}, {
      env: {
        GITHUB_TOKEN: 'undisclosed-github-token',
      },
    });
    const { fqRef } = await resp.json();
    assert.equal(resp.status, 200);
    assert.equal(fqRef, FULL_REF);
  });

  it('main() with path /_status_check/healthcheck.json reports status', async () => {
    const resp = await main({}, {}, {
      pathInfo: {
        suffix: '/_status_check/healthcheck.json',
      },
    });
    assert.equal(resp.status, 200);
    const result = await resp.json();

    assert.ok(result.github);
    delete result.github;
    assert.ok(result.response_time);
    delete result.response_time;
    delete result.process;
    assert.deepEqual(result, {
      status: 'OK',
      version: pkgJson.version,
    });
  });

  // eslint-disable-next-line func-names
  it('main function returns 502 for 5xx github server errors', async function () {
    const { server } = this.polly;
    server.host('https://github.com', () => {
      server.get('*').intercept((req, res) => {
        res.status(599);
      });
    });

    const { status } = await main({ owner: OWNER, repo: REPO, ref: SHORT_REF });
    assert.equal(status, 502);
  });
});

describe('network/server error tests', () => {
  it('main function returns 503 for network errors', async () => {
    // nock is also used by PollyJS under the hood.
    // In order to avoid unwanted side effects we have to reset nock.
    nock.cleanAll();
    nock.restore();
    nock.activate();

    // simulate network problem
    nock.disableNetConnect();
    try {
      const { status } = await main({ owner: OWNER, repo: REPO, ref: SHORT_REF });
      assert.equal(status, 503);
    } finally {
      nock.cleanAll();
      nock.enableNetConnect();
    }
  });
});


test/post-deploy.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
/* eslint-disable no-unused-expressions */

const chai = require('chai');
const chaiHttp = require('chai-http');
const packjson = require('../package.json');

chai.use(chaiHttp);
const { expect } = chai;

function getbaseurl() {
  const name = packjson.name.replace('@adobe/helix-', '');
  return `https://${name}-${process.env.CIRCLE_BRANCH || 'main'}.edgecompute.app`;
}

describe('Post-Deploy Tests #online #postdeploy', () => {
  it('correct sha is returned', async () => {
    let url;

    await chai
      .request(getbaseurl())
      .get(`?owner=adobe&repo=helix-resolve-git-ref&ref=v1.7.8`)
      .then((response) => {
        url = response.request.url;

        expect(response).to.have.status(200);
        expect(response).to.be.json;
      }).catch((e) => {
        e.message = `At ${url}\n      ${e.message}`;
        throw e;
      });
  }).timeout(10000);

  it('correct sha is returned when no ref has been set', async () => {
    let url;

    await chai
      .request(getbaseurl())
      .get(`?owner=trieloff&repo=test`)
      .then((response) => {
        url = response.request.url;

        expect(response).to.have.status(200);
        expect(response).to.be.json;
      }).catch((e) => {
        e.message = `At ${url}\n      ${e.message}`;
        throw e;
      });
  }).timeout(10000);
});


test/post-deploy.aws.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
/* eslint-disable no-unused-expressions */

const chai = require('chai');
const chaiHttp = require('chai-http');
const packjson = require('../package.json');
require('dotenv').config();

chai.use(chaiHttp);
const { expect } = chai;

function getbaseurl() {
  const package = 'helix-services';
  const name = packjson.name.replace('@adobe/helix-', '');
  let version = `${packjson.version}`;
  if (process.env.CI && process.env.CIRCLE_BUILD_NUM && process.env.CIRCLE_BRANCH !== 'main') {
    version = `ci${process.env.CIRCLE_BUILD_NUM}`;
  }
  return `/${package}/${name}_${version}`;
}

const HOST = `https://${process.env.HLX_AWS_API}.execute-api.${process.env.HLX_AWS_REGION}.amazonaws.com`;

describe('Post-Deploy Tests (AWS)', () => {
  before(function before() {
    if (!process.env.HLX_AWS_API || !process.env.HLX_AWS_REGION) {
      this.skip();
    }
  });

  it('correct sha is returned', async () => {
    let url;

    await chai
      .request(HOST)
      .get(`${getbaseurl()}?owner=adobe&repo=helix-resolve-git-ref&ref=v1.7.8`)
      .then((response) => {
        url = response.request.url;

        expect(response).to.have.status(200);
        expect(response).to.be.json;
      }).catch((e) => {
        e.message = `At ${url}\n      ${e.message}`;
        throw e;
      });
  }).timeout(10000);

  it('correct sha is returned when no ref has been set', async () => {
    let url;

    await chai
      .request(HOST)
      .get(`${getbaseurl()}?owner=trieloff&repo=test`)
      .then((response) => {
        url = response.request.url;

        expect(response).to.have.status(200);
        expect(response).to.be.json;
      }).catch((e) => {
        e.message = `At ${url}\n      ${e.message}`;
        throw e;
      });
  }).timeout(10000);
});


.eslintrc.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

module.exports = {
  root: true,
  parser: "@typescript-eslint/parser",
  plugins: [
    "@typescript-eslint",
  ],
  extends: [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
  ],
  parserOptions: {
    ecmaVersion: 2020,
    sourceType: "module",
    ecmaFeatures: {}
  },
  globals: {
    "BigInt64Array": "readonly",
    "BigUint64Array": "readonly",
    "__non_webpack_require__": "readonly"
  },

  // === General rules =========================================================

  rules: {
    // Omitted semicolons are hugely popular, yet within the compiler it makes
    // sense to be better safe than sorry.
    "semi": "error",

    // Our code bases uses 2 spaces for indentation, and we enforce it here so
    // files don't mix spaces, tabs or different indentation levels.
    "indent": ["error", 2, {
      "SwitchCase": 1,
      "VariableDeclarator": "first",
      "offsetTernaryExpressions": true,
      "ignoredNodes": [ // FIXME: something's odd here
        "ConditionalExpression > *",
        "ConditionalExpression > * > *",
        "ConditionalExpression > * > * > *"
      ]
    }],

    // This is mostly visual style, making comments look uniform.
    "spaced-comment": ["error", "always", {
      "markers": ["/"],   // triple-slash
      "exceptions": ["/"] // all slashes
    }],

    // This tends to be annoying as it encourages developers to make everything
    // that is never reassigned a 'const', sometimes semantically incorrect so,
    // typically leading to huge diffs in follow-up PRs modifying affected code.
    "prefer-const": "off",

    // It is perfectly fine to declare top-level variables with `var`, yet this
    // rule doesn't provide configuration options that would help.
    "no-var": "off",

    // Quite often, dealing with multiple related cases at once or otherwise
    // falling through is exactly the point of using a switch.
    "no-fallthrough": "off",

    // Typical false-positives here are `do { ... } while (true)` statements or
    // similar, but the only option provided here is not checking any loops.
    "no-constant-condition": ["error", { checkLoops: false }],

    // Functions are nested in blocks occasionally, and there haven't been any
    // problems with this so far, so turning the check off.
    "no-inner-declarations": "off",

    // Quite common in scenarios where an iteration starts at `current = this`.
    "@typescript-eslint/no-this-alias": "off",

    // Disabled here, but enabled again for JavaScript files.
    "no-unused-vars": "off",

    // Disabled here, but enabled again for TypeScript files.
    "@typescript-eslint/no-unused-vars": "off"
  },
  overrides: [

    // === JavaScript rules ====================================================

    {
      env: {
        "browser": true,
        "amd": true,
        "node": true,
        "es6": true
      },
      files: [
        "**/*.js",
        "bin/*"
      ],
      rules: {
        // We are testing both ESM and UMD, so don't limit us.
        "@typescript-eslint/no-var-requires": "off",

        // This rule does not behave well in JS files.
        "@typescript-eslint/explicit-module-boundary-types": "off",

        // Enforcing to remove function parameters on stubs makes code less
        // maintainable, so we instead allow unused function parameters.
        "no-unused-vars": [
          "warn", {
            "vars": "local",
            "args": "none",
            "ignoreRestSiblings": false
          }
        ]
      }
    },

    // === TypeScript rules ====================================================

    {
      files: [
        "**/*.ts"
      ],
      rules: {
        // Enforcing to remove function parameters on stubs makes code less
        // maintainable, so we instead allow unused function parameters.
        "@typescript-eslint/no-unused-vars": [
          "warn", {
            "vars": "local",
            "varsIgnorePattern": "^[A-Z](?:From|To)?$", // ignore type params
            "args": "none",
            "ignoreRestSiblings": false
          }
        ]
      }
    },

    // === AssemblyScript rules (extends TypeScript rules) =====================

    {
      files: [
        "fastly/assembly/**/*.ts",
      ],
      rules: {
        // Namespaces are quite useful in AssemblyScript
        "@typescript-eslint/no-namespace": "off",

        // There is actually codegen difference here
        "@typescript-eslint/no-array-constructor": "off",

        // Sometimes it can't be avoided to add a @ts-ignore
        "@typescript-eslint/ban-ts-comment": "off",

        // Utilized to achieve portability in some cases
        "@typescript-eslint/no-non-null-assertion": "off",
      }
    },
  ]
};

src/index.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

const { wrap: helixStatus } = require('@adobe/helix-status');
const { logger } = require('@adobe/helix-universal-logger');
const { wrap } = require('@adobe/helix-shared');
const { Response } = require('@adobe/helix-universal');
const lookup = require('./lookup');

async function main(req, context) {
  // TODO: query params are used a lot and should be context properties
  const url = new URL(req.url);
  const params = Array.from(url.searchParams.entries()).reduce((p, [key, value]) => {
    // eslint-disable-next-line no-param-reassign
    p[key] = value;
    return p;
  }, {});

  // augment token if missing
  if (!params.GITHUB_TOKEN) {
    params.GITHUB_TOKEN = context.env.GITHUB_TOKEN || req.headers.get('x-github-token');
  }
  const result = await lookup(params);
  return new Response(JSON.stringify(result.body), {
    headers: result.headers,
    status: result.statusCode,
  });
}

/**
 * Main function called by the openwhisk invoker.
 * @param params Action params
 * @returns {Promise<*>} The response
 */
module.exports.main = wrap(main)
  .with(helixStatus, { github: 'https://github.com/adobe/helix-resolve-git-ref.git/info/refs?service=git-upload-pack' })
  .with(logger.trace)
  .with(logger);


src/lookup.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-disable prefer-promise-reject-errors */

const {
  resolve: resolveRef,
  ResolveError,
  NetworkError,
} = require('@adobe/gh-resolve-ref');

/**
 * This is the main function. It resolves the specified reference to the corresponding
 * sha of the HEAD commit at `ref`.
 *
 * If the specified repository is private you have to provide a valid GitHub access token
 * either via `x-github-token` header or `GITHUB_TOKEN` action parameter.
 *
 * @param {Object} params The OpenWhisk parameters
 * @param {string} params.owner GitHub organization or user
 * @param {string} params.repo GitHub repository name
 * @param {string} [params.ref=<default branch>] git reference (branch or tag name)
 * @param {Object} params.__ow_headers The request headers of this web action invokation
 * @returns {Promise<object>} result
 * @returns {string} result.sha the sha of the HEAD commit at `ref`
 * @returns {string} result.fqRef the fully qualified name of `ref`
 *                                (e.g. `refs/heads/<branch>` or `refs/tags/<tag>`)
 */
function lookup(params) {
  const {
    owner,
    repo,
    ref,
    GITHUB_TOKEN: token,
  } = params;

  return resolveRef({
    owner,
    repo,
    ref,
    token,
  })
    .then((result) => {
      if (result) {
        return {
          statusCode: 200,
          headers: { 'Content-Type': 'application/json' },
          body: result,
        };
      } else {
        return {
          statusCode: 404,
          body: 'ref not found',
        };
      }
    })
    .catch((err) => {
      if (err instanceof TypeError) {
        return {
          statusCode: 400,
          body: 'owner and repo are mandatory parameters',
        };
      } else if (err instanceof NetworkError) {
        // (temporary?) network issue
        return {
          statusCode: 503, // service unavailable
          body: `failed to fetch git repo info: ${err.message}`,
        };
      } else /* istanbul ignore next */ if (err instanceof ResolveError) {
        const { statusCode, message } = err;
        let status = 500;
        if (statusCode >= 500 && statusCode <= 599) {
          // bad gateway
          status = 502;
        } else /* istanbul ignore next */ if (statusCode === 404) {
          // repo not found
          status = 404;
        }
        return {
          statusCode: status,
          body: `failed to fetch git repo info (statusCode: ${statusCode}, message: ${message})`,
        };
      } else {
        /* istanbul ignore next */
        return {
          statusCode: 500,
          body: `failed to fetch git repo info: ${err})`,
        };
      }
    });
}

module.exports = lookup;


fastly/assembly/index.ts
import { Request, Response, Fastly, Headers, ResponseInit } from "@fastly/as-compute";
import { CoralogixLogger } from "./coralogix";

function getQueryParam(qs: string, param: string): string {
  const pairs = qs.split("&");
  for (let i = 0; i < pairs.length; i++) {
    if (pairs[i].indexOf(param + "=")==0) {
      return pairs[i].substr(param.length + 1);
    }
  }
  return "";
}

function getQueryString(path: string):string {
  if (path.indexOf("?") > 0) {
    return path.substring(path.indexOf("?") + 1);
  }
  return "";
}

// The entry point for your application.
//
// Use this function to define your main request handling logic. It could be
// used to route based on the request properties (such as method or path), send
// the request to a backend, make completely new requests, and/or generate
// synthetic responses.
function main(req: Request): Response {
  const logger = new CoralogixLogger("helix-resolve-git-ref", req);


  // We can filter requests that have unexpected methods.
  const VALID_METHODS = ["HEAD", "GET", "POST"];
  if (!VALID_METHODS.includes(req.method())) {
    logger.error("Invalid method " + req.method());
    return new Response(String.UTF8.encode("This method is not allowed"), {
      status: (405 as i16),
    });
  }

  let urlParts = req.url().split("//").pop().split("/");
  let path = ("/" + urlParts.join("/"));
  let qs = getQueryString(path);

  const owner = getQueryParam(qs, "owner");
  const repo = getQueryParam(qs, "repo"); 
  let ref = getQueryParam(qs, "ref");
  let sha = "";

  if (owner != "" && repo != "" && true) {
    let cacheOverride = new Fastly.CacheOverride();
    cacheOverride.setTTL(30);

    const myreq = new Request("https://github.com/" + owner + "/" + repo + ".git/info/refs?service=git-upload-pack", {});

    const myresp = Fastly.fetch(myreq, {
      backend: "GitHub",
      cacheOverride,
    }).wait();

    logger.debug("Response received for " + owner + "/" + repo);

    if (myresp.status() >= 400 && myresp.status() < 500) {
      logger.error("Repo not found "  + owner + "/" + repo + "(" + myresp.status().toString(10) + ")");
      return new Response(String.UTF8.encode('failed to fetch git repo info (statusCode: ' + myresp.status().toString(10) +', statusMessage: ' + myresp.statusText() + ')'), {
        status: 404
      });
    }

    if (myresp.status() >= 500) {
      logger.error("Bad gateway "  + owner + "/" + repo + "(" + myresp.status().toString(10) + ")");
      return new Response(String.UTF8.encode('failed to fetch git repo info (statusCode: ' + myresp.status().toString(10) +', statusMessage: ' + myresp.statusText() + ')'), {
        status: 502 // bad gateway
      });
    }

    const lines = myresp.text().split("\n");
    for (let i = 0; i < lines.length; i++) {
      if (ref == "" && lines[i].indexOf("symref=HEAD:") > 0) {
        let refline = lines[i].substr(lines[i].indexOf("symref=HEAD:") + 23);
        ref = refline.substr(0, refline.indexOf(" "));
      }
      if (ref != "" && sha == "" && lines[i].indexOf(" refs/heads/" + ref) > 0) {
        sha = lines[i].substr(4, lines[i].indexOf(" refs/heads/" + ref) - 4);
        ref = "refs/heads/" + ref;
      }
      if (ref != "" && sha == "" && lines[i].indexOf(" refs/tags/" + ref) > 0) {
        sha = lines[i].substr(4, lines[i].indexOf(" refs/tags/" + ref) - 4);
        ref = "refs/tags/" + ref;
      }
    }


    if (sha == "") {
      let init = new ResponseInit();
      init.status = 404;
      return new Response(String.UTF8.encode('ref not found ' + ref), init);
    }

    const myheaders = new Headers();
    myheaders.set("Content-Type", "application/json");

    logger.debug("returning " + owner + "/" + repo + " " + sha);

    return new Response(String.UTF8.encode('{ "fqRef": "' + ref + '", "sha": "' + sha + '"  }'), {
      status: 200,
      headers: myheaders
    });
  }

  // Console.log('{ "timestamp": ' + Date.now().toString() + ', "applicationName":"fastly-edgecompute", "subsystemName":"helix-resolve-git-ref", "severity": 4, "json": { "message": "required parameters missing", "cdn": { "url": "' + req.url() + '" }}}');
  return new Response(String.UTF8.encode('owner and repo are mandatory parameters!'), {
    status: 400
  });
}

// Get the request from the client.
let req = Fastly.getClientRequest();

// Pass the request to the main request handler function.
let resp = main(req);

// Send the response back to the client.
Fastly.respondWith(resp);
